### PR TITLE
Copy avoidance in SWAP* insertion ranges

### PR DESCRIPTION
**Pull request by kkarbowiak**
## Issue

This is a draft of proposed change for #657.

The `prepostview` is a template class representing a lightweight view over contents of a vector (or a part of it) with an additional element in front or at back. It has just enough facilities to allow it being used in a range-based for loop.

This is work in progress and needs some polishing before it can be merged, but should work and allow to run tests and benchmarks.

Basically, we are trading a copy + trivial iteration over copied range for no copy but slightly slower iteration (each dereference contains some logic).

## Tasks

 - [ ] ...
 - [ ] Update `docs/API.md` (remove if irrelevant)
 - [ ] Update `CHANGELOG.md` (remove if irrelevant)
 - [ ] review

### CODE REVIEW

**Comment by jcoupey** [comment]
> I want to look at instance-level profiling/flamegraph to dig a bit deeper before drawing any conclusion.

So I finally got around to doing more thorough profiling. I've compared 176f57a9 (parent on master) and current HEAD of this PR at 463320b3 on several instances.

I've been using perf to monitor the share of `ls::compute_best_swap_star_choice` in the overall computation (this function takes a significant part in `ls::Operator::gain` computations). The change in number of samples and share is not representative of any significant improvement with the PR.

@kkarbowiak I guess my initial intuition from #657 was flawed, much as suggested by [your comment](https://github.com/VROOM-Project/vroom/issues/657#issuecomment-2066515373).

**Comment by jcoupey** [comment]
Closing here as the initial copy-avoidance approach does not seem relevant after all. I'll keep the related ticket open though because it contains other interesting ideas.

**Comment by jcoupey** [comment]
@kkarbowiak thanks for allowing to evaluate the idea with this draft!
